File: pt-1/3889.mirror-frequency-distance.go
package leetcode

//
// @lc app=leetcode id=3889 lang=golang
//
// [3889] Mirror Frequency Distance
//
// Notes
// Count characters, then compare mirrored lowercase letters a/z through m/n and
// mirrored digits 0/9 through 4/5. The answer is the sum of absolute frequency
// differences. A fixed byte-frequency array is enough in Go. Time: O(n). Space:
// O(1).
//
// @lc code=start

func MirrorFrequency3889(s string) int {
	frequency := [256]int{}
	for i := range s {
		frequency[s[i]]++
	}

	answer := 0
	for offset := 0; offset < 13; offset++ {
		left := byte('a' + offset)
		right := byte('z' - offset)
		answer += abs3889(frequency[left] - frequency[right])
	}
	for offset := 0; offset < 5; offset++ {
		left := byte('0' + offset)
		right := byte('9' - offset)
		answer += abs3889(frequency[left] - frequency[right])
	}
	return answer
}

func abs3889(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// @lc code=end

File: Google/3889.mirror-frequency-distance.java
/*
 * @lc app=leetcode id=3889 lang=java
 *
 * [3889] Mirror Frequency Distance
 *
 * Count characters, then compare mirrored lowercase letters a/z, b/y, ...
 * and mirrored digits 0/9, 1/8, ... . Sum absolute frequency differences.
 *
 * Time: O(|s| + alphabet). Space: O(alphabet).
 */

// @lc code=start
class Solution {
    public int mirrorFrequency(String s) {
        int[] freq = new int[128];
        for (int i = 0; i < s.length(); i++) {
            freq[s.charAt(i)]++;
        }

        int answer = 0;
        for (int offset = 0; offset < 13; offset++) {
            answer += Math.abs(freq['a' + offset] - freq['z' - offset]);
        }
        for (int offset = 0; offset < 5; offset++) {
            answer += Math.abs(freq['0' + offset] - freq['9' - offset]);
        }
        return answer;
    }
}
// @lc code=end
